core/backend/src/ElementTreeWalker.ts

```diff
@@ -450,17 +450,22 @@ export class ElementSubTreeDeleter extends ElementTreeTopDown {
 /** Deletes an element tree starting with the specified top element. The top element is also deleted. Uses ElementTreeDeleter.
  * @param iModel The iModel
  * @param topElement The parent of the sub-tree
+ * @param maxPasses Maximum number of passes to make when deleting definitions
  * @beta
  */
-export function deleteElementTree(iModel: IModelDb, topElement: Id64String): void {
-  const del = new ElementTreeDeleter(iModel);
-  del.deleteNormalElements(topElement);
-  del.deleteSpecialElements();
+export function deleteElementTree(iModel: IModelDb, topElement: Id64String, maxPasses: number = 5): void {
```

Comment:
Avoid piling on optional arguments. Something like this instead, keeping the original function with two arguments to avoid breaking existing callers, and adding an overload taking an `Args` object where new stuff can be added moving forward.
```ts
/** Arguments supplied to [[deleteElementTree]].
 * @beta
 */
export interface DeleteElementTreeArgs {
  /** The iModel containing the elements to delete. */
  iModel: IModelDb;
  /** The Id of the root element of the tree to delete. */
  topElement: Id64String;
  /** The maximum number of passes to make when deleting definition elements.
   * Default: 5
   */
  maxPasses?: number;
}
/** @beta */
export function deleteElementTree(iModel: IModelDb, topElement: Id64String): void;
/** @beta */
export function deleteElementTree(args: DeleteElementTreeArgs): void;
/** @internal */
export function deleteElementTree(arg0: DeleteElementTreeArgs | IModelDb, arg1?: IModelDb): void {
  let maxPasses = 5;
  let iModel: IModelDb;
  let topElement: Id64String;
  if (arg0 instanceof IModelDb) {
    assert(typeof arg1 === "string";
    iModel = arg0;
    topElement = arg1;
  } else {
    iModel = arg0.iModel
    topElement = arg0.topElement;
    maxPasses = arg0.maxPasses;
  }

  /* ... */
}
```

Reply:
> Do you have a way to detect that deletion will never succeed because of used definition elements, and bail before reaching maxPasses?

No way that I know of. That would come up if some of the definitions were used by elements outside of the tree parented by topElement. NativeDb.queryDefinitionElementUsagedoes not return information about where definitions are used. My code would have to essentially repeat the select statements that it executes and further analyze the results. Does not seem worth it to me.

Reply:
I added the overloads